### PR TITLE
PR-0000: Solution Design

### PR DESCRIPTION
**Pull request by radisvaliullin**
Design Document:
- Simple description of implementation approach.
- Contains Diagram of system components (packages).
- Short description of each components and them relations.

### CODE REVIEW

File: docs/design.md [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+    1. Periodically PING upstreams and mark them health/unhealthу.
+1. Runs balancer:
+    1. Based on the balancer algorithms (least) and healthiness provides the next upstream address for the handler.
+1. TCP server: 
+    1. accept connection and assign Handler.
+    1. Run Handler:
+        1. Handle TLS Handshake (using TLS Pkg). Get a TLS Session.
+        1. Requests UpStream Address from balancer.
+        1. Establish connection with UpStream.
+        1. Forward TLS Session to UpStream.
```
Comment (espadolini) [inline docs/design.md:26]: What do you mean by "forward TLS session to the upstream"?

File: docs/design.md [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+
+## General Diagram
+[<img src="design_diagram.png" width="480"/>](design_diagram.png)
+
+## Packages need to be created.
+1. TCP Server Package (Pkg) - provides TCP listener and connection accept.
+1. Conn Handler Pkg - makes TLS Handshake, upstream establishing, TLS Session forward to upstream, data forward.
+1. TLS Pkg - provides tools for TLS handshake and TLS session forward.
+1. Balancer Pkg - provides balancing. A least balancing, conn limit.
+1. HealthCheck Pkg - checks healthiness of upstreams.
```
Comment (espadolini) [inline docs/design.md:11]: How will the health check work?

File: docs/design.md [hunk 3]
```diff
@@ -0,0 +1,28 @@
+# Design of The Balancer Server
+
+## General Diagram
+[<img src="design_diagram.png" width="480"/>](design_diagram.png)
+
+## Packages need to be created.
+1. TCP Server Package (Pkg) - provides TCP listener and connection accept.
+1. Conn Handler Pkg - makes TLS Handshake, upstream establishing, TLS Session forward to upstream, data forward.
+1. TLS Pkg - provides tools for TLS handshake and TLS session forward.
+1. Balancer Pkg - provides balancing. A least balancing, conn limit.
```
Comment (espadolini) [inline docs/design.md:10]: Why does the connection limit have to be in the balancer, if it's supposed to be per-user?

**Review by espadolini — COMMENTED** [review]
There's no mention of how the authn scheme is going to work, or how the authz is involved in selecting which upstreams are available to different users; could you expand on that?

**Review by nklaassen — COMMENTED** [review]
To quote the challenge doc:
> The design document should cover: design approach, scope, proposed API, security considerations, CLI UX, and implementation details where appropriate.
>
> Develop a simple authorization scheme that defines what upstreams are available to which clients

What I mostly see here is a list of packages. I would like to see:
- some description of the library's API
- plan for authn/authz
- some example CLI UX of how someone will use this loadbalancer
- a bit of detail on how you will implement a least connections load balancer

File: docs/design.md (hunk 1 above)
Comment (radisvaliullin) [reply docs/design.md:26]: I planned just forward raw tcp stream from client connection to upstream. But I need handshake mTLS on balancer side for authn of user in balancer. I planned reproduce same handshake for upstream side and after just forward tcp stream. But looks like simplest way accept TLS connection and dial new tls connection with upstream (using same client cert). I will update design doc.

File: docs/design.md (hunk 2 above)
Comment (radisvaliullin) [reply docs/design.md:11]: I planned just periodically establish new tcp connection send "ping" message and get "pong" response.

File: docs/design.md (hunk 3 above)
Comment (radisvaliullin) [reply docs/design.md:10]: It might be other place.

**Comment by radisvaliullin** [comment]
@nklaassen I will update doc. I also missed authz package. I thought I need just pass mtls handshake, but I need handle clients (limit for client). I will fix that.

**Pull request closed without merging**